### ISSUE TITLE
my R hangs whenever I call the reticulate package?

### ISSUE THREAD

**Issue by mxomsibi** [rstudio/reticulate#415]
Hi guys,

My 'reticulate' package seem to have been working fine but started acting funny on the 25/26th of Desember. It hangs in there for a long dear life and never recovers, to only abort the R session. I tried closing all programs and restarting my laptop to no avail...

Even relaxed for a couple of days but today when I kicked it again it still does the same.

Please help

**Comment by harryprince** [comment]
it might be your program error, for example, print too much message in R notebook chunks, installing wrong python package and so on.

**Comment by tghoward** [comment]
I am having this problem too. Here are specifics in my case:
I am calling `arcpy` (which, granted, is a proprietary python package from ESRI). I can import it and make calls and everything works great. But as soon as I try to restart or close R or close RStudio, it all hangs and I must kill the process in task manager. 

details from a fresh start, with session info, below. 

```
R version 3.5.3 (2019-03-11) -- "Great Truth"
Copyright (C) 2019 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(reticulate)
> use_python("C:/Users/Tim_Howard/AppData/Local/ESRI/conda/envs/arcgispro-py3-clone", required = TRUE)
> arcpy <- import("arcpy")
> arcpy$CreateFileGDB_management("F:/tim/testArcpy", "test.gdb")
F:/tim/testArcpy\test.gdb
> sessionInfo()
R version 3.5.3 (2019-03-11)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows Server >= 2012 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] reticulate_1.10

loaded via a namespace (and not attached):
[1] compiler_3.5.3  Matrix_1.2-15   tools_3.5.3     Rcpp_1.0.0     
[5] grid_3.5.3      jsonlite_1.6    lattice_0.20-38
> 
```
Here's how it looks when it hangs (if attempting to close RStudio). 

![R_screenshot](https://user-images.githubusercontent.com/6129193/54839273-e4495400-4ca0-11e9-9cbe-0ed486c1dd88.jpg)

Is there a way I can force python to detach or disconnect or something?  Aside from this is works great!

**Comment by kevinushey** [comment]
> arcpy$CreateFileGDB_management("F:/tim/testArcpy", "test.gdb")

Does this command try to read input from the user?

**Comment by tghoward** [comment]
No, the command in this case is to create a new fileGeodatabase, which is a spatial geodatabase used by this GIS software.  The call succeeds. I need to run other commands too, but, as far as I can tell, once I run any `arcpy` command, that action causes R to hang when attempting to close it. It does not hang during use ... I can keep issuing commands and they are successful.

**Comment by kevinushey** [comment]
Interesting; thanks.

Would you be able to share a reproducible example? E.g. something I could copy + paste to run and reproduce the code locally (preferably without depending on any large data files)

**Comment by tghoward** [comment]
Yes, the three lines of code above is all you need (replacing the `test` paths and names with your own). 

But, ugh, you need an installed copy of esri's arcPro, which you can download a 30-day trial here: https://pro.arcgis.com/en/pro-app/get-started/install-and-sign-in-to-arcgis-pro.htm
And then use the python enviornment installed with it:
https://pro.arcgis.com/en/pro-app/arcpy/get-started/installing-python-for-arcgis-pro.htm

I realize this is asking a lot and don't necessarily expect you to go through all of this.  I'll try to research other possible ways to get into a test environment of arcpy. Perhaps there are others that have ArcGIS installed?

**Comment by tghoward** [comment]
@kevinushey thanks for engaging here and for your questions. Obviously if there is anything I can do here at this end to help debug I'd be more than happy to. AFAIK there isn't any other way to use ESRI's python library other than installing their software. So I understand that's probably a deal breaker. I came here because the package works just fine in Python directly (and I had spent many hours trying work-arounds on my own in `reticulate`). 

Again, thanks for thinking about this.

**Comment by kevinushey** [comment]
Debugging these sorts of issues (especially on Windows) can be quite challenging.

What I would end up doing if I were investigating this is trying to attach a debugger (e.g. GDB, WinDbg, the Visual Studio debugger) to the R session, and see if there's anything 'obvious' going on when the R session tries to close. For example, it might be getting deadlocked trying to run some clean-up code in the Python session, or it might stuck unexpectedly waiting for some kind of input.

Also worth asking: what happens if you're working in a 'regular' Python session? Does the session just close without anything 'special' happening? Or does it perhaps prompt the user with something like "Do you want to save your workspace? [y/n/c]: ".

**Comment by tghoward** [comment]
I just double checked in a regular session, and no, there isn't any user prompt on close. It behaves like any other package.   The one big thing unique about it, though, is that it requires a license and perhaps there's some kind of check-in or check-out process in the background that is getting stuck. 

Thanks for the tip - I'll try to look into debugging it.    The behavior was also why I was asking about some way to 'hard-disconnect' from python (or the package) or something like that. It doesn't really look like there is anything along those lines? For reticulate?

**Comment by tghoward** [comment]
Oooh, the plot thickens. R closes just fine when using R-Gui (Not RStudio). This code runs fine and R quits after the last line:
```
R version 3.5.1 (2018-07-02) -- "Feather Spray"
Copyright (C) 2018 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

Microsoft R Open 3.5.1
The enhanced R distribution from Microsoft
Microsoft packages Copyright (C) 2018 Microsoft Corporation

Using the Intel MKL for parallel mathematical computing (using 8 cores).

Default CRAN mirror snapshot taken on 2018-08-01.
See: https://mran.microsoft.com/.

> library(reticulate)
Warning message:
package ‘reticulate’ was built under R version 3.5.2 
> use_python("C:/Users/Tim_Howard/AppData/Local/ESRI/conda/envs/arcgispro-py3-clone", required = TRUE)
> arcpy <- import("arcpy")
> arcpy$CreateFileGDB_management("F:/tim/testArcpy", "test.gdb")
F:/tim/testArcpy\test.gdb
> q()
```
So it is an RStudio Thing. If that's helpful at all.   (note in this example I'm using MRO)

**Comment by dpphat** [comment]
I'm not sure if it's the same issue but when I am using the R console and I get the error:
Error in if (!is.na(arch)) { : argument is of length zero
but when I use Microsoft R Open the code runs to completion without error.

It seems the R console can't find my python distribution, even when I specify it using use_python(), but MRO can.

**Comment by mkoohafkan** [comment]
old thread, but I can confirm this is still happening and seems to be specific to ArcGIS. For me, the problem is reproducible after running any write operation (regardless of usage of a particular extension) but it is not caused by read operations or changing settings like reading environments etc.

You can check out the tests in https://github.com/mkoohafkan/arcpy, everything seems fine if you run the tests manually but the R session will hang once you try to quit after running one of the write operations (e.g., `arcpy$management$CopyFeatures()` or `arcpy$sa$CreateConstantRaster()`). R will therefore also hang when trying to run `devtools::check()` because it cannot quit the testing environment.

Is there a way to manually kill or rest a `reticulate` python environment?

**Comment by philiporlando** [comment]
Currently experiencing similar issues when incorporating `reticulate` and `arcpy` into an R package. My `test()` calls run without getting stuck or producing any errors. However, when I try to run `check()` it just hangs indefinitely on `Running 'testthat.R`...

![image](https://user-images.githubusercontent.com/24445174/151234187-c55f1593-a085-49e8-9566-08e18e2e5635.png)

![image](https://user-images.githubusercontent.com/24445174/151233107-315f747f-042f-4eb0-84f1-ea35f9c87461.png)

My guess is that there is lingering detritus within my test environment that is not getting cleaned before the next test... I have tried following [this solution](https://github.com/r-lib/testthat/issues/144), but I'm still running into the same issue and I think it is specific to `arcpy` or `reticulate`. Hoping someone else has figured out a solution by now, but let's be real, we're dealing with Esri software and I'm not very optimistic...
